### PR TITLE
types: exclude null and undefined from useResult pick fn and generics

### PR DESCRIPTION
**Pull request by ferm10n**
I was excited to learn about https://github.com/vuejs/vue-apollo/pull/1062 which fixed the potentially undefined result.value, but ran into an issue with the types in `useResult()`. Here's the example

```ts
const pickedResult = useResult(result, [], data => data.pickableThings);
```

To get the types to work, I have to provide values for the generics which is a bit verbose, but fine:

```ts
const pickedResult = useResult<Result, PickedResult, PickedResult>(result, [], data => data.pickableThings);
```

But now that the `result.value` is correctly typed, it reveals that the type of the [`data`](https://github.com/vuejs/vue-apollo/blob/v4/packages/vue-apollo-composable/src/useResult.ts#L63) argument in the pick function needs to be fixed. The pick function [will never be called](https://github.com/vuejs/vue-apollo/blob/v4/packages/vue-apollo-composable/src/useResult.ts#L77) if its value is undefined, so data should exclude undefined from its type. This PR is just a small change to implement that.

### CODE REVIEW

**Comment by ferm10n** [comment]
P.S. I still feel like I'm learning how to write good PR descriptions. I'd appreciate if there's any feedback on how I wrote this. Mainly concerned about being too wordy, or not giving enough info that's helpful.

**Comment by ferm10n** [comment]
I'm assuming the build failure isn't an issue since the target branch's build is still failing too for the same reason

**Comment by skix123** [comment]
Hi @ferm10n, thanks for this PR! I believe it solves this issue: https://github.com/vuejs/vue-apollo/issues/1231

Would love to see it merged

File: packages/vue-apollo-composable/src/useResult.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ export type UseResultReturn<T> = Readonly<Ref<Readonly<T>>>
  * @returns Readonly ref with `undefined` or the resolved `result`.
  */
 export function useResult<TResult, TResultKey extends keyof NonNullable<TResult> = keyof NonNullable<TResult>> (
-  result: Ref<TResult>
+  result: Ref<TResult | null | undefined>
```
Comment (Akryum) [inline packages/vue-apollo-composable/src/useResult.ts:20]: Why those changes?

**Comment by Akryum** [comment]
Thanks for the PR! I fixed the original issue in a different way in https://github.com/vuejs/vue-apollo/commit/4475805cb6dd48a1f0bdaa8df60432ebdc9963cd

**Comment by ferm10n** [comment]
Thanks for taking a look at this. I'm concerned with the DeepNonNullable type though. Isn't it possible for a query to return a null somewhere nested in the response?

I haven't verified this yet this is actually an issue yet though.

**Comment by Akryum** [comment]
That's what the try catch is for
